File: content/snippets/python/interview_lcci_03_02_min_stack.py
class MinStack:
    def __init__(self) -> None:
        self.values: list[int] = []
        self.mins: list[int] = []

    def push(self, x: int) -> None:
        self.values.append(x)
        if not self.mins or x <= self.mins[-1]:
            self.mins.append(x)

    def pop(self) -> None:
        if not self.values:
            return
        if self.values[-1] == self.mins[-1]:
            self.mins.pop()
        self.values.pop()

    def top(self) -> int:
        return self.values[-1] if self.values else -1

    def get_min(self) -> int:
        return self.mins[-1] if self.mins else -1


def main() -> None:
    st = MinStack()
    st.push(3)
    st.push(5)
    st.push(2)
    st.push(2)
    print(st.get_min())
    st.pop()
    print(st.get_min())
    st.pop()
    print(st.get_min())


if __name__ == "__main__":
    main()

File: content/snippets/cpp/interview_lcci_03_02_min_stack.cpp
#include <iostream>
#include <stack>

class MinStack {
public:
    void push(int x) {
        values.push(x);
        if (mins.empty() || x <= mins.top()) {
            mins.push(x);
        }
    }

    void pop() {
        if (values.empty()) return;
        if (values.top() == mins.top()) {
            mins.pop();
        }
        values.pop();
    }

    int top() const {
        return values.empty() ? -1 : values.top();
    }

    int getMin() const {
        return mins.empty() ? -1 : mins.top();
    }

private:
    std::stack<int> values;
    std::stack<int> mins;
};

int main() {
    MinStack st;
    st.push(3);
    st.push(5);
    st.push(2);
    st.push(2);
    std::cout << st.getMin() << '\n';
    st.pop();
    std::cout << st.getMin() << '\n';
    st.pop();
    std::cout << st.getMin() << '\n';
    return 0;
}
